homeassistant/components/mqtt/discovery.py

```diff
@@ -98,8 +104,22 @@
 
             _LOGGER.info("Found new component: %s %s", component, discovery_id)
 
-            await async_load_platform(
-                hass, component, platform, payload, hass_config)
+            if platform in CONFIG_ENTRY_PLATFORMS.get(component, []):
+                async_dispatcher_send(hass, 'mqtt_discovery_new_{}_{}'.format(
+                    component, platform), payload)
+            else:
+                await async_load_platform(
+                    hass, component, platform, payload, hass_config)
+
+    # TODO: Do we have to load all MQTT platforms on-demand or is this OK too?
```

Comment:
> not in hass.config.components:

So that check would just check if the platform, say `sensor.mqtt`, is loaded at all? But what if it is already loaded through `configuration.yaml`, but *not* through a config entry. Then we would still need to forward the config entry so that `async_setup_entry` is called.

Is there a check to see if a platform is already loaded **as a config entry**?

Reply:
> I also think that we could probably do with just 1 dispatch signal instead of having one for new and one for updated entries?

A problem is that the update dispatch call is sending the raw MQTT message (so that it can check if the device should be deleted) and this here is sending the parsed JSON message.

Probably this will need to be revised in the future anyway to a single dispatch call for both, but it seems a bit out of context for this PR. I could do it of course if you want.

Reply:
So since only discovery is forwarding config entries for MQTT, you can just keep track in here by storing a set in `hass.data`